src/transforms/sample/config.rs

```diff
@@ -0,0 +1,144 @@
+    /// and the rate is applied to the buckets themselves to sample `1/N` buckets.  The overall rate
+    /// of sampling may differ from the configured one if values in the field are not uniformly
+    /// distributed. If left unspecified, or if the event doesn’t have `key_field`, then the
+    /// event is sampled independently.
+    ///
+    /// This can be useful to, for example, ensure that all logs for a given transaction are
+    /// sampled together, but that overall `1/N` transactions are sampled.
+    #[configurable(metadata(docs::examples = "message",))]
+    pub key_field: Option<String>,
+
+    /// A logical condition used to exclude events from sampling.
+    pub exclude: Option<AnyCondition>,
+}
+
+impl GenerateConfig for SampleConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            rate: 10,
+            key_field: None,
+            exclude: None::<AnyCondition>,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "sample")]
+impl TransformConfig for SampleConfig {
+    async fn build(&self, context: &TransformContext) -> crate::Result<Transform> {
+        Ok(Transform::function(Sample::new(
+            Self::NAME.to_string(),
+            self.rate,
+            self.key_field.clone(),
+            self.exclude
+                .as_ref()
+                .map(|condition| condition.build(&context.enrichment_tables))
+                .transpose()?,
+        )))
+    }
+
+    fn input(&self) -> Input {
+        Input::new(DataType::Log | DataType::Trace)
+    }
+
+    fn outputs(
+        &self,
+        _: vector_lib::enrichment::TableRegistry,
+        input_definitions: &[(OutputId, schema::Definition)],
+        _: LogNamespace,
+    ) -> Vec<TransformOutput> {
+        vec![TransformOutput::new(
+            DataType::Log | DataType::Trace,
+            input_definitions
+                .iter()
+                .map(|(output, definition)| {
+                    (
+                        output.clone(),
+                        definition.clone().with_source_metadata(
+                            SampleConfig::NAME,
+                            Some(LegacyKey::Overwrite(owned_value_path!("sample_rate"))),
+                            &owned_value_path!("sample_rate"),
+                            Kind::bytes(),
+                            None,
+                        ),
+                    )
+                })
+                .collect(),
+        )]
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        event::LogEvent, test_util::components::assert_transform_compliance,
+        transforms::test::create_topology,
+    };
+    use tokio::sync::mpsc;
+    use tokio_stream::wrappers::ReceiverStream;
+
+    #[test]
+    fn generate_config() {
+        crate::test_util::test_generate_config::<SampleConfig>();
+    }
+
+    #[tokio::test]
```

Comment:
I would expect this test to be in `transform.rs`. Any reason for moving it here?

Reply:
This unit test instantiates a `SampleConfig`. I kept tests that relied on the config structure here so that there  was a clear separation with the modules and feature deps.